openshift/helper/hashes.py

```diff
@@ -0,0 +1,50 @@
+# Implement ConfigMapHash and SecretHash equivalents
+# Based on https://github.com/kubernetes/kubernetes/pull/49961
+
+import json
+import hashlib
+
+try:
+    import string
+    maketrans = string.maketrans
+except AttributeError:
+    maketrans = str.maketrans
+
+try:
+    from collections import OrderedDict
+except ImportError:
+    from orderreddict import OrderedDict
+
+
+def sorted_dict(unsorted_dict):
+    result = OrderedDict()
+    for (k, v) in sorted(unsorted_dict.items()):
+        if isinstance(v, dict):
+            v = sorted_dict(v)
+        result[k] = v
+    return result
+
+
+def generate_hash(resource):
+    # Get name from metadata
+    resource['name'] = resource.get('metadata', {}).get('name', '')
```

Comment:
What's the purpose of moving name to the top level and then deleting it?

Reply:
it's basically to leave `resource` the same as before the call (I could use deepcopy instead) and the hash is based on the `data`, `kind` and `name` keys of the resource - adding `name` as a top level key makes the `marshal` function a lot simpler. It's not the only possible implementation though.